### PR TITLE
v0.0.10

### PR DESCRIPTION
**Pull request by cameronmartino**
v0.0.10 (2023-12-01)

### Bug fixes

* Removed np.int types
    * fixes bug that prevents use in newer versions of numpy / QIIME2
        * see issue #71

### Features [experimental]

* rpca projection of new data and with cross validation
    * A new function for RPCA where new data can be projected into an existing ordination.
    * Also allows for internal CV, where the hold out data is projected into the training ordination, which is then used to re-build the test data, and the error between the projection and the real test data is calculated.
        * see issue #70
* tempted.py and assocated tests/commands/tutorials
    * Added in a python implementation TEMPTED, more details on the methods inm this paper [here](https://www.biorxiv.org/content/10.1101/2023.07.26.550749v1).
* joint-rpca and associated tests/commands/tutorials
    * Joint-RPCA is an extention of RPCA for multiple omics types.
* qc_rarefaction
    * This function compares the mantel correlation of distances to the abs. differences in sample sum between rarefied and unrarefied input data. This is an easy check to ensure the results are not bieng significantly altered by non-rarefaction is cases of large differences (e.g., low-biomass) or where the sample sum differences can not be seperated from the phenotypes/low-rank structure of the data (i.e., deep sequencing of controls and shallow of sick).
        * see issue #70

### Deprecated functionality

* auto_rpca across the package and rank_estimate in optspace.py
    * was not performing correctly and no reasonable quick fix is available.
        * see issue #70

### CODE REVIEW

**Comment by SilasK** [comment]
Looking forward to the new version.

**Comment by SilasK** [comment]
I'm testing this branch and have some comments.

Don't forget to include in the release notes the breaking change from 

from `auto_rpca` to `gemelli.rpca import rpca`

Using simulated random data I get the 

```
gemelli/gemelli/preprocessing.py:426: RuntimeWarning: divide by zero encountered in log
  mat = np.log(matrix_closure(mat))
  ```

**Comment by cameronmartino** [comment]
Thanks, @SilasK! That warning can be ignored, it is meant to be suppressed within Gemelli but if numpy is re-imported before running RPCA it will get un-suppressed. Not sure how to fix that or if I want to since the user might want that warning for other parts of their own code. It comes from the RCLR command where the infs/-infs/NaNs are later ignored. If you want to suppress it in your scripts add 

```python
numpy.seterr(divide = 'ignore')
```
after your numpy import. 

Let me know if you run into any other issues/comments on the new branch!

**Comment by SilasK** [comment]
Maby there is a divide function that can handle nans.

**Comment by SilasK** [comment]
Otherwise I could use it.

**Comment by gibsramen** [comment]
Self-assigning for review

**Comment by gibsramen** [comment]
I have *no* idea why, but on my end the Joint-RPCA CLI notebook fails at the PERMANOVA step.

image

I managed to fix this with the following change but it's unclear why it even failed.

image

Pandas version is `1.5.3`

**Comment by cameronmartino** [comment]
@SilasK looking into it seems like there are some issues around that ([here](https://github.com/numpy/numpy/issues/18744)). That may be why we opted for the way we do it, I don't remember. I opened an issue for it and will look more deeply into this for a future version. Thanks for pointing it out!

**Comment by cameronmartino** [comment]
@gibsramen Fixed, thanks!

**Review by gibsramen — COMMENTED** [review]
Couple changes requested, thanks!

File: gemelli/tempted.py [hunk 1]
```diff
@@ -0,0 +1,1340 @@
...
+                               DEFAULT_TEMPTED_PC,
+                               DEFAULT_TEMPTED_EP,
+                               DEFAULT_TEMPTED_SMTH,
+                               DEFAULT_TEMPTED_RES,
+                               DEFAULT_TEMPTED_MAXITER,
+                               DEFAULT_TEMPTED_RH as DEFAULT_TRH,
+                               DEFAULT_TEMPTED_RHC as DEFAULT_RC,
+                               DEFAULT_TEMPTED_SVDC,
+                               DEFAULT_TEMPTED_SVDCN as DEFAULT_TSCN)
+np.random.seed(42)
```
[the remark this answers is not in the document]
Comment (cameronmartino) [reply gemelli/tempted.py]: @gibsramen Good point! This was unnecessary anyway, it is removed now.

File: gemelli/preprocessing.py [hunk 2]
```diff
@@ -10,15 +10,21 @@
...
 from biom import Table
 from skbio import TreeNode
 from .base import _BaseConstruct
+from inspect import getfullargspec
 from gemelli._defaults import DEFAULT_MTD
+from skbio.stats.composition import clr
 from skbio.diversity._util import _vectorize_counts_and_tree
 from bp import parse_newick, to_skbio_treenode
+from scipy.sparse.linalg import svds
+np.random.seed(42)
```
[the remark this answers is not in the document]
Comment (cameronmartino) [reply gemelli/preprocessing.py]: @gibsramen Good point! This was unnecessary anyway, it is removed now.

File: gemelli/optspace.py [hunk 3]
```diff
@@ -212,6 +197,162 @@ def solve(self, obs):
         U, S, V = svd_sort(U, S, V)
         return U, S, V
 
+    def joint_solve(self, multiple_obs):
```
[the remark this answers is not in the document]
Comment (cameronmartino) [reply gemelli/optspace.py:200]: Done! Thanks!!

**Pull request merged**